juju/client/gocookies.py

```diff
@@ -0,0 +1,105 @@
+import datetime
+import json
+import http.cookiejar as cookiejar
+import time
+import pyrfc3339
+
+
+class GoCookieJar(cookiejar.FileCookieJar):
+    '''A CookieJar implementation that reads and writes cookies
+    to the cookiejar format as understood by the Go package
+    github.com/juju/persistent-cookiejar.'''
+    def _really_load(self, f, filename, ignore_discard, ignore_expires):
+        '''Implement the _really_load method called by FileCookieJar
+        to implement the actual cookie loading'''
+        now = time.time()
+        '''This method is called by FileCookieJar to do the cookie loading'''
+        for c in map(_new_py_cookie, json.loads(f.read())):
+            print('loading cookie {}'.format(c))
+            if not ignore_expires and c.is_expired(now):
+                continue
+            self.set_cookie(c)
+
+    def save(self, filename=None, ignore_discard=False, ignore_expires=False):
+        '''Implement the FileCookieJar abstract method'''
+        if filename is None:
+            if self.filename is not None:
+                filename = self.filename
+            else:
+                raise ValueError(cookiejar.MISSING_FILENAME_TEXT)
+
+        # TODO: obtain file lock, read contents of file, and merge with
+        # current content.
+        with open(filename, "w") as f:
+            now = time.time()
+            go_cookies = []
+            for cookie in self:
+                if not ignore_discard and cookie.discard:
+                    continue
+                if not ignore_expires and cookie.is_expired(now):
+                    continue
+                go_cookies.append(_new_go_cookie(cookie))
+            f.write(json.dumps(go_cookies))
+
+
+def _new_py_cookie(go_cookie):
+    '''Convert a Go-style JSON-unmarshaled cookie
+    into a Python cookie'''
+    c = go_cookie
+    expires = None
+    if c.get('Expires') is not None:
+        t = pyrfc3339.parse(c['Expires'])
+        expires = t.strftime("%s")
+    return cookiejar.Cookie(
+        version=0,
+        name=c['Name'],
+        value=c['Value'],
+        port=None,
```

Comment:
Shall we do something for the port ?

Reply:
I don't understand why ports are there at all - the cookie RFC doesn't provide support for port-specific cookies.